File: v2/206.py



# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        # 6 star, pass, 必须熟练掌握
        # 就是前一个节点，当前节点和下一个节点之间的变换关系
        node = head
        prev = None
        while node:
            next = node.next
            node.next = prev
            prev = node
            node = next
        return prev

File: go/206.go

//package main
//
//import "fmt"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// type ListNode struct {
// 	Val int
// 	Next *ListNode
//}
// 4 star,
func reverseList(head *ListNode) *ListNode {
	node := head
	var pre, next *ListNode
	for node != nil {
		next = node.Next
		node.Next = pre
		pre = node
		node = next
	}
	return pre

}

//func main()  {
//	a := &ListNode{1, nil}
//	b := &ListNode{2, nil}
//	c := &ListNode{3, nil}
//	a.Next = b
//	b.Next = c
//	d := reverseList(a)
//	fmt.Println(d)
//
//}
